be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsWithMapHandler.scala

```diff
@@ -0,0 +1,140 @@
+
+      }
+  })
+
+  private def heartbeatHandler(dbActorRef: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map({
+    case Right(jsonRpcMessage: JsonRpcRequest) =>
+      /** first step is to retrieve the received heartbeat from the jsonRpcRequest * */
+      val receivedHeartBeat: Map[Channel, Set[Hash]] = jsonRpcMessage.getParams.asInstanceOf[Heartbeat].channelsToMessageIds
+
+      /** second step is to retrieve the local set of channels * */
+      var setOfChannels: Set[Channel] = Set()
+      val ask = dbActorRef ? DbActor.GetAllChannels()
+      Await.ready(ask, duration).value match {
+        case Some(Success(DbActor.DbActorGetAllChannelsAck(channels))) => setOfChannels = channels
+        case Some(Failure(ex: DbActorNAckException))                   => Left(PipelineError(ex.code, s"couldn't retrieve local set of channels", jsonRpcMessage.getId))
+        case reply                                                     => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, s"heartbeatHandler failed : unexpected DbActor reply '$reply'", jsonRpcMessage.getId))
+      }
+
+      /** third step is to ask the DB for the content of each channel in terms of message ids. * */
+      val localHeartBeat: mutable.HashMap[Channel, Set[Hash]] = mutable.HashMap()
+      setOfChannels.foreach(channel => {
+        val ask = dbActorRef ? DbActor.ReadChannelData(channel)
+        Await.ready(ask, duration).value match {
+          case Some(Success(DbActor.DbActorReadChannelDataAck(channelData))) =>
+            val setOfIds = channelData.messages.toSet
+            localHeartBeat += (channel -> setOfIds)
+          case Some(Failure(ex: DbActorNAckException)) => Left(PipelineError(ex.code, s"couldn't readChannelData for local heartbeat", jsonRpcMessage.getId))
+          case reply                                   => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, s"heartbeatHandler failed : unexpected DbActor reply '$reply'", jsonRpcMessage.getId))
+        }
+      })
+
+      /** finally, we only keep from the received heartbeat the message ids that are not contained in the locally extracted heartbeat. * */
+      val missingIds: mutable.HashMap[Channel, Set[Hash]] = mutable.HashMap()
+      receivedHeartBeat.keys.foreach(channel => {
+        if (localHeartBeat.contains(channel)) {
+          missingIds += (channel -> receivedHeartBeat(channel).diff(localHeartBeat(channel)))
+        }
+      })
+      Right(JsonRpcRequest(RpcValidator.JSON_RPC_VERSION, MethodType.GET_MESSAGES_BY_ID, GetMessagesById(missingIds.toMap), None))
+
+    case Right(jsonRpcMessage: JsonRpcResponse) =>
+      Left(PipelineError(ErrorCodes.SERVER_ERROR.id, "HeartbeatHandler received a 'JsonRpcResponse'", jsonRpcMessage.id))
+    case graphMessage @ _ => graphMessage
+  }).filter(!isEmptyAnswer(_))
+
+  private def getMessagesByIdHandler(dbActorRef: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map({
```

Comment:
No need for opening parenthesis in the map
```suggestion
  private def getMessagesByIdHandler(dbActorRef: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
```